storage/src/storage/sqlite/sync.rs

```diff
@@ -0,0 +1,923 @@
+            .map_err(Into::into)
+    }
+
+    fn get_block_state(&mut self, hash: &Digest) -> Result<BlockStatus> {
+        let output: Option<Option<usize>> = self
+            .conn
+            .query_row(r"SELECT canon_height FROM blocks WHERE hash = ?", [hash], |row| {
+                row.get(0)
+            })
+            .optional()?;
+
+        Ok(match output {
+            None => BlockStatus::Unknown,
+            Some(None) => BlockStatus::Uncommitted,
+            Some(Some(n)) => BlockStatus::Committed(n),
+        })
+    }
+
+    fn get_block_states(&mut self, hashes: &[Digest]) -> Result<Vec<BlockStatus>> {
+        let mut out = Vec::with_capacity(hashes.len());
+        for hash in hashes {
+            let state = self.get_block_state(hash)?;
+            out.push(state);
+        }
+        Ok(out)
+        // let dyn_hashes: Vec<&dyn ToSql> = hashes.iter().map(|x| -> &dyn ToSql { x }).collect::<Vec<_>>();
+        // let mut stmt = self.conn.prepare_cached("SELECT hash, canon_height FROM blocks WHERE hash IN rarray(?)")?;
+        // let mut output = IndexMap::<Digest, BlockStatus>::new();
+        // for hash in hashes {
+        //     output.insert(hash.clone(), BlockStatus::Unknown);
+        // }
+        // for row in stmt.query_map(&dyn_hashes[..], |row| Ok((row.get(0)?, row.get(1)?)))? {
+        //     let (hash, canon_height): (Digest, Option<usize>) = row?;
+        //     *output.get_mut(&hash).unwrap() = match canon_height {
+        //         Some(height) => BlockStatus::Committed(height),
+        //         None => BlockStatus::Uncommitted,
+        //     };
+        // }
+        // Ok(output.into_iter().map(|x| x.1).collect())
+    }
+
+    fn get_block(&mut self, hash: &Digest) -> Result<SerialBlock> {
+        let header = self.get_block_header(hash)?;
+        let mut stmt = self.conn.prepare_cached(
+            "SELECT
+            transactions.transaction_id,
+            network,
+            ledger_digest,
+            old_serial_number1,
+            old_serial_number2,
+            new_commitment1,
+            new_commitment2,
+            program_commitment,
+            local_data_root,
+            value_balance,
+            signature1,
+            signature2,
+            new_record1,
+            new_record2,
+            transactions.proof,
+            memo,
+            inner_circuit_id
+        FROM transactions
+        INNER JOIN transaction_blocks on transaction_blocks.transaction_id = transactions.id
+        INNER JOIN blocks on blocks.id = transaction_blocks.block_id
+        WHERE blocks.hash = ?
+        ORDER BY transaction_blocks.block_order ASC",
+        )?;
+        let rows = stmt.query_map([hash], |row| {
+            Ok(SerialTransaction {
+                id: read_static_blob(row, 0)?,
+                network: Network::from_id(row.get(1)?),
+                ledger_digest: row.get(2)?,
+                old_serial_numbers: vec![row.get(3)?, row.get(4)?],
```

Comment:
[applies to new_commitments, signatures and new_records too] do we expect there to ever be more than 2 entries for these items? if not, we could use 2-elem arrays instead of vectors to avoid 4 allocations for each transaction

Reply:
we don't need to hyperoptimize vectors so much. that being said -- i highly doubt we will ever have any other than 2, unless we change our transaction-record model (2 in, 2 out). it's nice to keep the floor open to whatever snarkvm throws at us though, sqlite doesn't have variable length arrays generally (i think there is an extension), so i just took some liberty here to make it work.